specs/custom-iresourcemanager-spec.md

```diff
@@ -0,0 +1,109 @@
+Providing a custom MRT Core `IResourceManager` for WinUI 3 to use
+===
+
+# Background
+
+The WinUI 3 framework instantiates an MRT Core
+[`ResourceManager`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources.resourcemanager?view=windows-app-sdk-1.2)
+in order to resolve resource URIs. While this default `ResourceManager` is generally sufficient, 
+there are some scenarios in which an app needs non-standard behavior in order to resolve a
+particular resource URI. In order to address this requirement, this document describes a new
+API that would allow an app to provide its own custom implementation of the
+[`IResourceManager`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources.iresourcemanager?view=windows-app-sdk-1.2)
+for WinUI 3 to use in lieu of its standard `ResourceManager`.
+
+# API Pages
+
+_(Each of the following L2 sections correspond to a page that will be on docs.microsoft.com)_
+
+## Application.ResourceManagerRequested Event
+
+Raised during startup of a new WinUI thread to give the app a chance to provide its own
+[`IResourceManager`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources.iresourcemanager?view=windows-app-sdk-1.2)
+implementation to be used by the framework for resolving resource URIs.
+
+_Spec note: is there a version of the 'ResourceDictionary and XAML resource references' article that 
+links to the Windows App SDK documentation for relevant APIs?_
+
+```c#
+public event TypedEventHandler<Application, ResourceManagerRequestedEventArgs> 
+ResourceManagerRequested
+
+```
+
+### Remarks
+
+It is strongly recommended that apps which need to listen for this event register their handler in their `App` 
+class's constructor so that it is available during initial app launch. This event is raised once per WinUI
+thread during initialization. Because the framework assumes that each `IResourceManager` instance returned
```

Comment:
> This event is raised once per WinUI thread during initialization

This scared me, because it made me think it gets raised during thread initialization (and how do I register my event handler before the initial process thread starts running?). But really, it is raised during WinUI initialization. Maybe

> This event is raised once per WinUI thread during WinUI initialization.

(How does a thread initialize WinUI anyway? I assume there's some API call to say "Okay, WinUI, please initialize yourself on this thread.")

Reply:
I don't know the APIs off the top of my head as I'm only familiar with them in passing, but basically creating a XAML Island on a new thread. IIRC from a discussion with @MikeHillberg a while ago this is still considered an experimental feature so my goal with the statement was to be accurate while avoiding any specific details as to how multiple WinUI threads might come into existence.

I feel like somebody who is unfamiliar with our architecture would read "WinUI initialization" to mean a once-per-process-lifetime thing. What I really wanted to say here is "once per WinUI *core* during initialization" but, again, that requires knowledge of our architecture that isn't well-documented for outsiders. I do see your point, though; I just don't think "WinUI initialization" sufficiently clarifies the semantics but neither do I have any other ideas that come to mind.